### PR TITLE
Remote Graph Wrappers for cuGraph-Service

### PR DESCRIPTION
**Pull request by alexbarghi-nv**
Resolves https://github.com/rapidsai/graph_dl/issues/76
Merge after #2804 

This PR defines a remote interface for `Graph` and `PropertyGraph` so they can be easily substituted in a client application that does not use `cugraph` directly.  For `PropertyGraph`, this PR also implements most of the functionality of a native `PropertyGraph`. 

Note: this PR does not include the upcoming "dispatch" updates for running cugraph algorithms or subgraph extraction, which are covered under https://github.com/rapidsai/graph_dl/issues/80

### CODE REVIEW

File: python/cugraph_service/cugraph_service_client/remote_graph.py [hunk 1]
```diff
@@ -0,0 +1,530 @@
...
+            May be a list, or dictionary corresponding to column names.
+        """
+
+        if backend == "cupy":
+            if isinstance(data, np.ndarray):
+                data = cupy.array(data)
+            return data
+        else:
+            # cudf
+            df = cudf.DataFrame.from_records(data, columns=column_names)
```
Comment (VibhuJawa) [inline python/cugraph_service/cugraph_service_client/remote_graph.py:122]: I think if we are transforming 2d device arrays to cuDF we should do this via `dlpack`  and not via `from_records` . I wonder how this will work if the underlying array is `Row-major`  .

File: python/cugraph_service/cugraph_service_client/remote_graph.py [hunk 2]
```diff
@@ -0,0 +1,530 @@
...
+                for col_name, t in dtypes.items():
+                    df[col_name] = df[col_name].astype(t)
+            return df
+        # TODO support torch
+
+    @property
+    def _vertex_categorical_dtype(self):
+        if self.__vertex_categorical_dtype is None:
+            cats = self.vertex_types
+            self.__vertex_categorical_dtype = cudf.CategoricalDtype(cats)
```
Comment (VibhuJawa) [inline python/cugraph_service/cugraph_service_client/remote_graph.py:137]: This imposes a hard dependency of `cudf` being installed , I hope thats intensional.

File: python/cugraph_service/cugraph_service_client/remote_graph.py [hunk 3]
```diff
@@ -0,0 +1,530 @@
...
+        -------
+        A copy of df with additional columns corresponding to properties for
+        the edge in the row.
+        FIXME: also provide the ability to annotate vertex data.
+
+        Examples
+        --------
+        >>>
+        """
+        raise NotImplementedError("not ipmlemented")
```
Comment (VibhuJawa) [inline python/cugraph_service/cugraph_service_client/remote_graph.py:514]:
```suggestion
        raise NotImplementedError("not implemented")
```

Comment (VibhuJawa) [inline python/cugraph_service/cugraph_service_client/remote_graph.py:514]:
```suggestion
        raise NotImplementedError("not implemented")
```

File: python/cugraph_service/cugraph_service_server/cugraph_handler.py [hunk 4]
```diff
@@ -441,6 +442,9 @@ def load_csv_as_edge_data(
         if len(names) == 0:
             names = None
 
+        if edge_id_col_name == "":
+            edge_id_col_name = None
```
Comment (VibhuJawa) [inline python/cugraph_service/cugraph_service_server/cugraph_handler.py:446]: Why this change in default ?

**Review by VibhuJawa — CHANGES_REQUESTED** [review]
Thanks for working on this @alexbarghi-nv , I can see how this can be very useful. Overall approach looks good but have raised some questions.

File: python/cugraph/cugraph/structure/property_graph.py [hunk 5]
```diff
@@ -262,9 +262,13 @@ def get_num_vertices(self, type=None, *, include_edge_data=True):
             vert_sers = self.__get_all_vertices_series()
             if vert_sers:
                 if self.__series_type is cudf.Series:
-                    self.__num_vertices = cudf.concat(vert_sers).nunique()
+                    self.__num_vertices = cudf.concat(
+                        vert_sers, ignore_index=True
+                    ).nunique()
```
Comment (VibhuJawa) [inline python/cugraph/cugraph/structure/property_graph.py:267]: Can you merge in master as i think these changes are now In.

File: python/cugraph_service/cugraph_service_client/remote_graph.py (hunk 2 above)
Comment (alexbarghi-nv) [reply python/cugraph_service/cugraph_service_client/remote_graph.py:137]: not intentional; I will fix this

Comment (alexbarghi-nv) [reply python/cugraph_service/cugraph_service_client/remote_graph.py:137]: not intentional, I will fix it

File: python/cugraph_service/cugraph_service_client/remote_graph.py (hunk 1 above)
Comment (alexbarghi-nv) [reply python/cugraph_service/cugraph_service_client/remote_graph.py]: resolved; this method no longer handles incoming device arrays

File: python/cugraph_service/cugraph_service_client/remote_graph.py (hunk 2 above)
Comment (alexbarghi-nv) [reply python/cugraph_service/cugraph_service_client/remote_graph.py]: should be fixed now; backend is fixed in the remote property graph and cudf is an optional import

File: python/cugraph_service/cugraph_service_client/remote_graph.py [hunk 6]
```diff
@@ -0,0 +1,530 @@
...
+
+class RemotePropertyGraph:
+    # column name constants used in internal DataFrames
+    vertex_col_name = "_VERTEX_"
+    src_col_name = "_SRC_"
+    dst_col_name = "_DST_"
+    type_col_name = "_TYPE_"
+    edge_id_col_name = "_EDGE_ID_"
+    weight_col_name = "_WEIGHT_"
+    _default_type_name = ""
```
[the remark this answers is not in the document]
Comment (alexbarghi-nv) [reply python/cugraph_service/cugraph_service_client/remote_graph.py:195]: resolved; reliance on cudf is removed

File: python/cugraph_service/cugraph_service_server/cugraph_handler.py (hunk 4 above)
Comment (alexbarghi-nv) [reply python/cugraph_service/cugraph_service_server/cugraph_handler.py:455]: So that an empty string is not passed to `add_edge_data`, which will result in an error when the expected behavior is to auto-generate the ids.  The correct interpretation of "" received from the client is None, since "" is used as a substitute for None as `edge_id_col_name` has to be a string.

File: python/cugraph_service/cugraph_service_client/remote_graph.py [hunk 7]
```diff
@@ -0,0 +1,649 @@
...
+    else:
+        raise ValueError("dtypes must be None, a list/tuple, or a dict")
+
+    if not isinstance(data, np.ndarray):
+        raise TypeError("Numpy ndarray expected")
+
+    if backend == "cupy":
+        return [cupy.array(data[:, c], dtype=dtypes[c]) for c in range(data.shape[1])]
+    elif backend == "numpy":
+        print("dtypes:", dtypes)
```
Comment (VibhuJawa) [inline python/cugraph_service/cugraph_service_client/remote_graph.py:108]: Remove print

**Review by VibhuJawa — CHANGES_REQUESTED** [review]
Have requested small changes, looks good other-wise

File: python/cugraph/cugraph/structure/property_graph.py [hunk 8]
```diff
@@ -1334,6 +1334,11 @@ def __get_all_vertices_series(self):
         if epd is not None:
             vert_sers.append(epd[self.src_col_name])
             vert_sers.append(epd[self.dst_col_name])
+        if len(vert_sers) > 1 and not all(
+            cudf.api.types.is_dtype_equal(vert_sers[0].index.dtype, s.index.dtype)
+            for s in vert_sers
+        ):
+            vert_sers = [s.reset_index(drop=True) for s in vert_sers]
```
[the remark this answers is not in the document]
Comment (alexbarghi-nv) [reply python/cugraph/cugraph/structure/property_graph.py:1341]: It's not needed.  For some reason this gets added back whenever I merge in branch-22.12.

**Review by rlratzel — APPROVED** [review]

**Comment by VibhuJawa** [comment]
@alexbarghi-nv  is this good to merge ?

**Comment by alexbarghi-nv** [comment]
This is good to merge.

**Comment by VibhuJawa** [comment]
@gpucibot merge

**Comment by VibhuJawa** [comment]
@rlratzel , Can you merge this please, i dont think i have privileges.

**Comment by rlratzel** [comment]
oops, sorry for the delay, didn't notice it was ready.

@gpucibot merge

**Comment by rlratzel** [comment]
@gpucibot merge
